File: iam-login-service/src/main/webapp/resources/iam/js/service/passwordreset.service.js
/*
 * Copyright (c) Istituto Nazionale di Fisica Nucleare (INFN). 2016-2019
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict'

angular.module('passwordResetApp').factory('ResetPasswordService', ResetPasswordService);

ResetPasswordService.$inject = ['$http', '$httpParamSerializerJQLike'];

function ResetPasswordService($http, $httpParamSerializerJQLike){
	
	var service = {
		forgotPassword : forgotPassword,
		changePassword : changePassword,
	};
	
	return service;
	
	function forgotPassword(email){
		
		var data = $httpParamSerializerJQLike({
				email: email
		});
		
		var config = {
			headers : {
				'Content-Type' : 'application/x-www-form-urlencoded'
			}
		} 
		
		return $http.post('/iam/password-reset/token', data, config);
		
	};
	
	function changePassword(resetKey, newPassword){
		var data = $httpParamSerializerJQLike({
			token : resetKey,
			password : newPassword
		});
		
		var config = {
			headers : {
				'Content-Type': 'application/x-www-form-urlencoded'
			}
		}
		
		return $http.post('/iam/password-reset', data, config);
	}
}

File: iam-login-service/src/main/java/it/infn/mw/iam/api/account/password_reset/PasswordResetService.java
/**
 * Copyright (c) Istituto Nazionale di Fisica Nucleare (INFN). 2016-2019
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package it.infn.mw.iam.api.account.password_reset;

import it.infn.mw.iam.api.account.password_reset.error.BadUserPasswordError;
import it.infn.mw.iam.api.account.password_reset.error.InvalidPasswordResetTokenError;
import it.infn.mw.iam.api.account.password_reset.error.UserNotActiveOrNotVerified;

/**
 * 
 * The IAM password reset service
 *
 */
public interface PasswordResetService {


  /**
   * Validates a password reset token.
   * 
   * @param resetToken the password reset token to be validated
   * 
   * @throws InvalidPasswordResetTokenError if the password reset token is not valid
   */
  public void validateResetToken(String resetToken);

  /**
   * Resets the password for an account, given a valid password reset token
   * 
   * @param resetToken the password reset token
   * 
   * @param password the password to be set
   * 
   * @throws InvalidPasswordResetTokenError if the password reset token is not valid
   */
  public void resetPassword(String resetToken, String password);

  /**
   * Creates a password reset token for the account linked with the email passed as argument.
   * 
   * @param email the email linked to the account for which the password must be reset
   */
  public void createPasswordResetToken(String email);

  /**
   * Updates the password for the account identified by both username and the current active password
   *
   * @param username the account username
   *
   * @param oldPassword the current active password
   *
   * @param newPassword the password to be set
   *
   * @throws UserNotActiveOrNotVerified if the user is not enabled
   *
   * @throws BadUserPasswordError if the @oldPassword doesn't match
   */
  public void updatePassword(String username, String oldPassword, String newPassword);

}
